File: python/BestTimeToBuyAndSellStockIII.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        if not prices:
            return 0
        
        length = len(prices)
        minValue = prices[0]
        maxValue = prices[length - 1]
        left = [0] * length
        right = [0] * length
        
        for i in range(1, length):
            minValue = min(minValue, prices[i])
            left[i] = max(left[i - 1], prices[i] - minValue)
            
        for i in range(length - 1, 0, -1):
            maxValue = max(maxValue, prices[i])
            right[i - 1] = max(right[i], maxValue - prices[i - 1])
            
        res = 0
        for i in range(0, length):
            res = max(res, left[i] + right[i])
        return res

File: java/BestTimeToBuyAndSellStockIII.java
class Solution {
    public int maxProfit(int[] prices) {
        if (prices == null || prices.length < 2){
            return 0;
        }
        
        int length = prices.length;
        
        int res = 0;
        int min = prices[0];
        int max = prices[length - 1];
        int[] left = new int[length];
        int[] right = new int[length];
        
        for (int i = 1; i < length; i++){
            min = Math.min(min, prices[i]);
            left[i] = Math.max(left[i - 1], prices[i] - min);
        }
        
        for (int i = length - 1; i >= 1; i--){
            max = Math.max(max, prices[i]);
            right[i - 1] = Math.max(right[i], max - prices[i - 1]);
        }
        
        for (int i = 0; i < length; i++){
            res = Math.max(res, left[i] + right[i]);
        }
        
        return res;
    }
}
